File: coding-patterns/16-binary-search/118.py
# Number Range (Medium)

class Solution:
  def findRange(self, arr, key):
    result = [- 1, -1]
    result[0] = self.binary_search(arr, key, False)
    if result[0] != -1:
      result[1] = self.binary_search(arr, key, True)
    return result


  # modified Binary Search
  def binary_search(self, arr, key, findMaxIndex):
    keyIndex = -1
    start, end = 0, len(arr) - 1
    while start <= end:
      mid = start + (end - start) // 2
      if key < arr[mid]:
        end = mid - 1
      elif key > arr[mid]:
        start = mid + 1
      else: 
        keyIndex = mid
        if findMaxIndex:
          start = mid + 1  
        else:
          end = mid - 1 

    return keyIndex

File: coding-patterns/16-binary-search/118.cpp
// Number Range (Medium)

using namespace std;

#include <iostream>
#include <vector>

class Solution {
public:
  vector<int> findRange(const vector<int> &arr, int key) {
    vector<int> result(2, -1);
    result[0] = search(arr, key, false);

    if (result[0] != -1)
        result[1] = search(arr, key, true);

    return result;
  }

private:
    static int search(const vector<int> &arr, int key, bool maxid) {
        int keyid = -1;
        int start = 0;
        int end = arr.size() - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (key < arr[mid])
                end = mid - 1;
            else if (key > arr[mid])
                start = mid + 1;
            else {
                keyid = mid;
                if (maxid)
                    start = mid + 1;
                else 
                    end = mid - 1;
            }
        }

        return keyid;
    }

};
